File: course-template/tools/setup-verification.js
#!/usr/bin/env node

/**
 * Course Template Setup Verification Script
 * 
 * This script verifies that the student's development environment
 * is properly configured for the course.
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const os = require('os');

// ANSI color codes for console output
const colors = {
    green: '\x1b[32m',
    red: '\x1b[31m',
    yellow: '\x1b[33m',
    blue: '\x1b[34m',
    cyan: '\x1b[36m',
    reset: '\x1b[0m',
    bold: '\x1b[1m'
};

// Helper functions
const log = {
    success: (msg) => console.log(`${colors.green}✅ ${msg}${colors.reset}`),
    error: (msg) => console.log(`${colors.red}❌ ${msg}${colors.reset}`),
    warning: (msg) => console.log(`${colors.yellow}⚠️  ${msg}${colors.reset}`),
    info: (msg) => console.log(`${colors.blue}ℹ️  ${msg}${colors.reset}`),
    header: (msg) => console.log(`${colors.bold}${colors.cyan}\n=== ${msg} ===${colors.reset}`)
};

class SetupVerifier {
    constructor() {
        this.checks = [];
        this.passed = 0;
        this.failed = 0;
        this.warnings = 0;
        this.results = [];
    }

    async runAllChecks() {
        log.header('Course Template Setup Verification');
        console.log(`Platform: ${os.platform()} ${os.arch()}`);
        console.log(`Node.js Version: ${process.version}`);
        console.log(`Timestamp: ${new Date().toISOString()}\n`);

        // System checks
        await this.checkNodeVersion();
        await this.checkNpmVersion();
        await this.checkGitInstallation();
        await this.checkCodeEditor();
        
        // Project structure checks
        await this.checkProjectStructure();
        await this.checkDependencies();
        await this.checkEnvironmentFiles();
        
        // Optional tool checks
        await this.checkOptionalTools();
        
        // Network and permissions
        await this.checkNetworkAccess();
        await this.checkFilePermissions();
        
        // Generate summary
        this.generateSummary();
    }

    async checkNodeVersion() {
        log.header('Node.js Version Check');
        try {
            const version = process.version;
            const majorVersion = parseInt(version.slice(1).split('.')[0]);
            
            if (majorVersion >= 16) {
                log.success(`Node.js ${version} - Compatible`);
                this.recordResult('Node.js Version', 'PASS', `Version ${version} meets requirements (>=16)`);
            } else {
                log.error(`Node.js ${version} - Too old! Please upgrade to Node.js 16 or higher`);
                this.recordResult('Node.js Version', 'FAIL', `Version ${version} is too old. Requires >=16`);
            }
        } catch (error) {
            log.error('Node.js not found or not accessible');
            this.recordResult('Node.js Version', 'FAIL', 'Node.js not found');
        }
    }

    async checkNpmVersion() {
        log.header('npm Version Check');
        try {
            const npmVersion = execSync('npm --version', { encoding: 'utf8' }).trim();
            const majorVersion = parseInt(npmVersion.split('.')[0]);
            
            if (majorVersion >= 7) {
                log.success(`npm ${npmVersion} - Compatible`);
                this.recordResult('npm Version', 'PASS', `Version ${npmVersion} meets requirements (>=7)`);
            } else {
                log.warning(`npm ${npmVersion} - Consider upgrading to npm 7+`);
                this.recordResult('npm Version', 'WARNING', `Version ${npmVersion} works but upgrade recommended`);
            }
        } catch (error) {
            log.error('npm not found or not accessible');
            this.recordResult('npm Version', 'FAIL', 'npm not found');
        }
    }

    async checkGitInstallation() {
        log.header('Git Installation Check');
        try {
            const gitVersion = execSync('git --version', { encoding: 'utf8' }).trim();
            log.success(`${gitVersion} - Installed`);
            
            // Check git configuration
            try {
                const userName = execSync('git config user.name', { encoding: 'utf8' }).trim();
                const userEmail = execSync('git config user.email', { encoding: 'utf8' }).trim();
                
                if (userName && userEmail) {
                    log.success(`Git configured with: ${userName} <${userEmail}>`);
                    this.recordResult('Git Configuration', 'PASS', `Configured as ${userName} <${userEmail}>`);
                } else {
                    log.warning('Git not configured. Run: git config --global user.name "Your Name" && git config --global user.email "your.email@example.com"');
                    this.recordResult('Git Configuration', 'WARNING', 'Git not configured');
                }
            } catch (configError) {
                log.warning('Git not configured globally');
                this.recordResult('Git Configuration', 'WARNING', 'Not configured');
            }
            
            this.recordResult('Git Installation', 'PASS', gitVersion);
        } catch (error) {
            log.error('Git not found. Please install Git from https://git-scm.com/');
            this.recordResult('Git Installation', 'FAIL', 'Git not found');
        }
    }

    async checkCodeEditor() {
        log.header('Code Editor Check');
        const editors = [
            { name: 'VS Code', command: 'code --version', recommended: true },
            { name: 'Visual Studio', command: 'devenv /?', recommended: false },
            { name: 'Sublime Text', command: 'subl --version', recommended: false },
            { name: 'Atom', command: 'atom --version', recommended: false }
        ];

        let editorFound = false;
        for (const editor of editors) {
            try {
                execSync(editor.command, { stdio: 'ignore' });
                const status = editor.recommended ? '(Recommended)' : '';
                log.success(`${editor.name} found ${status}`);
                this.recordResult('Code Editor', 'PASS', `${editor.name} available`);
                editorFound = true;
                break;
            } catch (error) {
                // Editor not found, continue checking
            }
        }

        if (!editorFound) {
            log.warning('No recognized code editor found. We recommend VS Code: https://code.visualstudio.com/');
            this.recordResult('Code Editor', 'WARNING', 'No editor detected');
        }
    }

    async checkProjectStructure() {
        log.header('Project Structure Check');
        const requiredDirectories = [
            'course-template',
            'course-template/01-fundamentals',
            'course-template/02-intermediate',
            'course-template/03-advanced',
            'course-template/04-expert',
            'course-template/common',
            'course-template/templates',
            'course-template/assessments',
            'course-template/tools'
        ];

        const requiredFiles = [
            'course-template/README.md',
            'course-template/common/code-examples/javascript-starter.js',
            'course-template/common/code-examples/python-starter.py',
            'course-template/templates/lesson-template/README.md',
            'course-template/templates/exercise-template/README.md',
            'course-template/templates/project-template/README.md'
        ];

        let structureValid = true;

        // Check directories
        for (const dir of requiredDirectories) {
            if (fs.existsSync(dir) && fs.statSync(dir).isDirectory()) {
                log.success(`Directory: ${dir}`);
            } else {
                log.error(`Missing directory: ${dir}`);
                structureValid = false;
            }
        }

        // Check files
        for (const file of requiredFiles) {
            if (fs.existsSync(file) && fs.statSync(file).isFile()) {
                log.success(`File: ${file}`);
            } else {
                log.error(`Missing file: ${file}`);
                structureValid = false;
            }
        }

        this.recordResult('Project Structure', structureValid ? 'PASS' : 'FAIL', 
            structureValid ? 'All required files and directories present' : 'Missing required files/directories');
    }

    async checkDependencies() {
        log.header('Dependencies Check');
        
        // Check if package.json exists in course-template
        const packageJsonPath = path.join('course-template', 'package.json');
        
        if (fs.existsSync(packageJsonPath)) {
            try {
                const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
                log.success('package.json found');
                
                // Check if node_modules exists
                const nodeModulesPath = path.join('course-template', 'node_modules');
                if (fs.existsSync(nodeModulesPath)) {
                    log.success('node_modules directory found');
                    this.recordResult('Dependencies', 'PASS', 'Dependencies installed');
                } else {
                    log.warning('node_modules not found. Run: cd course-template && npm install');
                    this.recordResult('Dependencies', 'WARNING', 'Dependencies not installed');
                }
            } catch (error) {
                log.error('Invalid package.json format');
                this.recordResult('Dependencies', 'FAIL', 'Invalid package.json');
            }
        } else {
            log.info('No package.json found (optional for this template)');
            this.recordResult('Dependencies', 'INFO', 'No package.json required');
        }
    }

    async checkEnvironmentFiles() {
        log.header('Environment Files Check');
        
        const envFiles = [
            { path: 'course-template/.env.example', required: false, description: 'Environment template' },
            { path: 'course-template/.gitignore', required: true, description: 'Git ignore file' }
        ];

        for (const envFile of envFiles) {
            if (fs.existsSync(envFile.path)) {
                log.success(`${envFile.description}: ${envFile.path}`);
            } else if (envFile.required) {
                log.error(`Missing required file: ${envFile.path}`);
                this.recordResult('Environment Files', 'FAIL', `Missing ${envFile.path}`);
                continue;
            } else {
                log.info(`Optional file not found: ${envFile.path}`);
            }
        }

        this.recordResult('Environment Files', 'PASS', 'Environment setup acceptable');
    }

    async checkOptionalTools() {
        log.header('Optional Tools Check');
        
        const optionalTools = [
            { name: 'Python', command: 'python --version', alt: 'python3 --version' },
            { name: 'Java', command: 'java -version' },
            { name: 'Docker', command: 'docker --version' },
            { name: 'Yarn', command: 'yarn --version' }
        ];

        for (const tool of optionalTools) {
            try {
                const version = execSync(tool.command, { encoding: 'utf8', stdio: 'pipe' });
                log.success(`${tool.name}: Available`);
            } catch (error) {
                if (tool.alt) {
                    try {
                        const version = execSync(tool.alt, { encoding: 'utf8', stdio: 'pipe' });
                        log.success(`${tool.name}: Available (via ${tool.alt})`);
                        continue;
                    } catch (altError) {
                        // Fall through to not found
                    }
                }
                log.info(`${tool.name}: Not found (optional)`);
            }
        }

        this.recordResult('Optional Tools', 'INFO', 'Optional tools checked');
    }

    async checkNetworkAccess() {
        log.header('Network Access Check');
        
        const testUrls = [
            'https://registry.npmjs.org/',
            'https://github.com',
            'https://api.github.com'
        ];

        let networkIssues = false;
        for (const url of testUrls) {
            try {
                // Simple network check using Node.js built-ins
                const https = require('https');
                await new Promise((resolve, reject) => {
                    const req = https.get(url, { timeout: 5000 }, (res) => {
                        log.success(`Network access to ${url}: OK`);
                        resolve();
                    });
                    req.on('error', reject);
                    req.on('timeout', () => reject(new Error('Timeout')));
                });
            } catch (error) {
                log.warning(`Network access to ${url}: Failed (${error.message})`);
                networkIssues = true;
            }
        }

        this.recordResult('Network Access', networkIssues ? 'WARNING' : 'PASS', 
            networkIssues ? 'Some network issues detected' : 'Network access OK');
    }

    async checkFilePermissions() {
        log.header('File Permissions Check');
        
        try {
            // Test write permissions in current directory
            const testFile = path.join(process.cwd(), '.setup-test-' + Date.now());
            fs.writeFileSync(testFile, 'test');
            fs.unlinkSync(testFile);
            
            log.success('File write permissions: OK');
            this.recordResult('File Permissions', 'PASS', 'Write permissions available');
        } catch (error) {
            log.error(`File permission issue: ${error.message}`);
            this.recordResult('File Permissions', 'FAIL', `Permission error: ${error.message}`);
        }
    }

    recordResult(category, status, details) {
        this.results.push({ category, status, details });
        
        switch (status) {
            case 'PASS':
                this.passed++;
                break;
            case 'FAIL':
                this.failed++;
                break;
            case 'WARNING':
                this.warnings++;
                break;
        }
    }

    generateSummary() {
        log.header('Setup Verification Summary');
        
        console.log(`${colors.green}✅ Passed: ${this.passed}${colors.reset}`);
        console.log(`${colors.red}❌ Failed: ${this.failed}${colors.reset}`);
        console.log(`${colors.yellow}⚠️  Warnings: ${this.warnings}${colors.reset}`);
        
        if (this.failed === 0) {
            log.success('🎉 Congratulations! Your development environment is ready for the course.');
        } else {
            log.error('⚠️  Please fix the failed checks before starting the course.');
        }

        // Generate detailed report
        console.log('\n' + colors.bold + 'Detailed Results:' + colors.reset);
        console.log('='.repeat(80));
        
        for (const result of this.results) {
            const statusColor = result.status === 'PASS' ? colors.green : 
                              result.status === 'FAIL' ? colors.red : colors.yellow;
            
            console.log(`${statusColor}${result.status.padEnd(8)}${colors.reset} ${result.category.padEnd(20)} ${result.details}`);
        }

        // Save report to file
        this.saveReport();
        
        console.log('\n' + colors.cyan + 'Next Steps:' + colors.reset);
        if (this.failed > 0) {
            console.log('1. Fix the failed checks listed above');
            console.log('2. Re-run this verification script');
            console.log('3. Contact your instructor if you need help');
        } else {
            console.log('1. Start with Module 01 - Fundamentals');
            console.log('2. Read the course README.md for detailed instructions');
            console.log('3. Join the course discussion forum');
        }
        
        console.log('\n' + colors.blue + 'Support Resources:' + colors.reset);
        console.log('- Course Documentation: README.md');
        console.log('- Setup Guide: docs/setup.md');
        console.log('- Discussion Forum: [course-forum-url]');
        console.log('- Instructor Email: [instructor-email]');
    }

    saveReport() {
        const report = {
            timestamp: new Date().toISOString(),
            platform: `${os.platform()} ${os.arch()}`,
            nodeVersion: process.version,
            summary: {
                passed: this.passed,
                failed: this.failed,
                warnings: this.warnings
            },
            results: this.results
        };

        try {
            fs.writeFileSync('setup-verification-report.json', JSON.stringify(report, null, 2));
            log.info('📄 Detailed report saved to: setup-verification-report.json');
        } catch (error) {
            log.warning('Could not save detailed report file');
        }
    }
}

// Main execution
async function main() {
    const verifier = new SetupVerifier();
    
    try {
        await verifier.runAllChecks();
        
        // Exit with appropriate code
        process.exit(verifier.failed > 0 ? 1 : 0);
    } catch (error) {
        log.error(`Verification script error: ${error.message}`);
        process.exit(1);
    }
}

// Handle command line arguments
const args = process.argv.slice(2);
if (args.includes('--help') || args.includes('-h')) {
    console.log(`
Course Template Setup Verification

Usage: node setup-verification.js [options]

Options:
  --help, -h     Show this help message
  --version, -v  Show version information
  --verbose      Enable verbose output
  --json         Output results in JSON format

This script verifies that your development environment is properly
configured for the software development course.

Examples:
  node setup-verification.js
  node setup-verification.js --verbose
  node setup-verification.js --json > report.json
`);
    process.exit(0);
}

if (args.includes('--version') || args.includes('-v')) {
    console.log('Course Template Setup Verification v1.0.0');
    process.exit(0);
}

// Run the main function
if (require.main === module) {
    main();
}

module.exports = SetupVerifier;

File: course-template/tools/setup-verification.py
#!/usr/bin/env python3

"""
Course Template Setup Verification Script (Python Version)

This script verifies that the student's development environment
is properly configured for the course.
"""

import os
import sys
import json
import subprocess
import platform
import shutil
from datetime import datetime
from pathlib import Path
from typing import List, Dict, Tuple, Optional

# ANSI color codes for console output
class Colors:
    GREEN = '\033[32m'
    RED = '\033[31m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    CYAN = '\033[36m'
    RESET = '\033[0m'
    BOLD = '\033[1m'

class Logger:
    @staticmethod
    def success(msg: str) -> None:
        print(f"{Colors.GREEN}✅ {msg}{Colors.RESET}")
    
    @staticmethod
    def error(msg: str) -> None:
        print(f"{Colors.RED}❌ {msg}{Colors.RESET}")
    
    @staticmethod
    def warning(msg: str) -> None:
        print(f"{Colors.YELLOW}⚠️  {msg}{Colors.RESET}")
    
    @staticmethod
    def info(msg: str) -> None:
        print(f"{Colors.BLUE}ℹ️  {msg}{Colors.RESET}")
    
    @staticmethod
    def header(msg: str) -> None:
        print(f"{Colors.BOLD}{Colors.CYAN}\n=== {msg} ==={Colors.RESET}")

class SetupVerifier:
    """Main class for verifying course setup requirements"""
    
    def __init__(self):
        self.results: List[Dict[str, str]] = []
        self.passed = 0
        self.failed = 0
        self.warnings = 0
    
    async def run_all_checks(self) -> None:
        """Run all verification checks"""
        Logger.header('Course Template Setup Verification (Python)')
        print(f"Platform: {platform.system()} {platform.machine()}")
        print(f"Python Version: {sys.version}")
        print(f"Timestamp: {datetime.now().isoformat()}\n")
        
        # System checks
        await self.check_python_version()
        await self.check_pip_installation()
        await self.check_git_installation()
        await self.check_code_editor()
        
        # Project structure checks
        await self.check_project_structure()
        await self.check_python_dependencies()
        await self.check_environment_files()
        
        # Optional tool checks
        await self.check_optional_tools()
        
        # Network and permissions
        await self.check_network_access()
        await self.check_file_permissions()
        
        # Generate summary
        self.generate_summary()
    
    async def check_python_version(self) -> None:
        """Check if Python version meets requirements"""
        Logger.header('Python Version Check')
        
        version_info = sys.version_info
        version_str = f"{version_info.major}.{version_info.minor}.{version_info.micro}"
        
        if version_info >= (3, 8):
            Logger.success(f"Python {version_str} - Compatible")
            self.record_result('Python Version', 'PASS', 
                             f"Version {version_str} meets requirements (>=3.8)")
        else:
            Logger.error(f"Python {version_str} - Too old! Please upgrade to Python 3.8 or higher")
            self.record_result('Python Version', 'FAIL', 
                             f"Version {version_str} is too old. Requires >=3.8")
    
    async def check_pip_installation(self) -> None:
        """Check if pip is installed and working"""
        Logger.header('pip Installation Check')
        
        try:
            result = subprocess.run([sys.executable, '-m', 'pip', '--version'], 
                                  capture_output=True, text=True, check=True)
            pip_version = result.stdout.strip()
            Logger.success(f"{pip_version} - Available")
            self.record_result('pip Installation', 'PASS', pip_version)
        except (subprocess.CalledProcessError, FileNotFoundError):
            Logger.error('pip not found or not working')
            self.record_result('pip Installation', 'FAIL', 'pip not available')
    
    async def check_git_installation(self) -> None:
        """Check Git installation and configuration"""
        Logger.header('Git Installation Check')
        
        try:
            result = subprocess.run(['git', '--version'], capture_output=True, text=True, check=True)
            git_version = result.stdout.strip()
            Logger.success(f"{git_version} - Installed")
            
            # Check git configuration
            try:
                user_name = subprocess.run(['git', 'config', 'user.name'], 
                                         capture_output=True, text=True, check=True).stdout.strip()
                user_email = subprocess.run(['git', 'config', 'user.email'], 
                                          capture_output=True, text=True, check=True).stdout.strip()
                
                if user_name and user_email:
                    Logger.success(f"Git configured with: {user_name} <{user_email}>")
                    self.record_result('Git Configuration', 'PASS', 
                                     f"Configured as {user_name} <{user_email}>")
                else:
                    Logger.warning('Git not configured. Run: git config --global user.name "Your Name" && git config --global user.email "your.email@example.com"')
                    self.record_result('Git Configuration', 'WARNING', 'Git not configured')
            except subprocess.CalledProcessError:
                Logger.warning('Git not configured globally')
                self.record_result('Git Configuration', 'WARNING', 'Not configured')
            
            self.record_result('Git Installation', 'PASS', git_version)
        except (subprocess.CalledProcessError, FileNotFoundError):
            Logger.error('Git not found. Please install Git from https://git-scm.com/')
            self.record_result('Git Installation', 'FAIL', 'Git not found')
    
    async def check_code_editor(self) -> None:
        """Check for available code editors"""
        Logger.header('Code Editor Check')
        
        editors = [
            {'name': 'VS Code', 'command': ['code', '--version'], 'recommended': True},
            {'name': 'PyCharm', 'command': ['pycharm', '--version'], 'recommended': True},
            {'name': 'Sublime Text', 'command': ['subl', '--version'], 'recommended': False},
            {'name': 'Vim', 'command': ['vim', '--version'], 'recommended': False},
            {'name': 'Nano', 'command': ['nano', '--version'], 'recommended': False}
        ]
        
        editor_found = False
        for editor in editors:
            try:
                subprocess.run(editor['command'], capture_output=True, check=True)
                status = '(Recommended)' if editor['recommended'] else ''
                Logger.success(f"{editor['name']} found {status}")
                self.record_result('Code Editor', 'PASS', f"{editor['name']} available")
                editor_found = True
                break
            except (subprocess.CalledProcessError, FileNotFoundError):
                continue
        
        if not editor_found:
            Logger.warning('No recognized code editor found. We recommend VS Code or PyCharm')
            self.record_result('Code Editor', 'WARNING', 'No editor detected')
    
    async def check_project_structure(self) -> None:
        """Check if required project structure exists"""
        Logger.header('Project Structure Check')
        
        required_directories = [
            'course-template',
            'course-template/01-fundamentals',
            'course-template/02-intermediate',
            'course-template/03-advanced',
            'course-template/04-expert',
            'course-template/common',
            'course-template/templates',
            'course-template/assessments',
            'course-template/tools'
        ]
        
        required_files = [
            'course-template/README.md',
            'course-template/common/code-examples/javascript-starter.js',
            'course-template/common/code-examples/python-starter.py',
            'course-template/templates/lesson-template/README.md',
            'course-template/templates/exercise-template/README.md',
            'course-template/templates/project-template/README.md'
        ]
        
        structure_valid = True
        
        # Check directories
        for directory in required_directories:
            if Path(directory).is_dir():
                Logger.success(f"Directory: {directory}")
            else:
                Logger.error(f"Missing directory: {directory}")
                structure_valid = False
        
        # Check files
        for file_path in required_files:
            if Path(file_path).is_file():
                Logger.success(f"File: {file_path}")
            else:
                Logger.error(f"Missing file: {file_path}")
                structure_valid = False
        
        self.record_result('Project Structure', 'PASS' if structure_valid else 'FAIL',
                          'All required files and directories present' if structure_valid 
                          else 'Missing required files/directories')
    
    async def check_python_dependencies(self) -> None:
        """Check Python dependencies and virtual environment"""
        Logger.header('Python Dependencies Check')
        
        # Check if requirements.txt exists
        requirements_path = Path('course-template/requirements.txt')
        
        if requirements_path.exists():
            Logger.success('requirements.txt found')
            
            # Check if virtual environment is activated
            if hasattr(sys, 'real_prefix') or (hasattr(sys, 'base_prefix') and sys.base_prefix != sys.prefix):
                Logger.success('Virtual environment detected')
                self.record_result('Virtual Environment', 'PASS', 'Virtual environment active')
            else:
                Logger.warning('No virtual environment detected. Consider using: python -m venv venv && source venv/bin/activate')
                self.record_result('Virtual Environment', 'WARNING', 'No virtual environment')
            
            # Try to check installed packages
            try:
                with open(requirements_path, 'r') as f:
                    requirements = [line.strip() for line in f if line.strip() and not line.startswith('#')]
                
                missing_packages = []
                for requirement in requirements:
                    package_name = requirement.split('==')[0].split('>=')[0].split('<=')[0]
                    try:
                        __import__(package_name)
                    except ImportError:
                        missing_packages.append(package_name)
                
                if missing_packages:
                    Logger.warning(f'Missing packages: {", ".join(missing_packages)}. Run: pip install -r requirements.txt')
                    self.record_result('Python Dependencies', 'WARNING', f'Missing: {", ".join(missing_packages)}')
                else:
                    Logger.success('All required packages appear to be installed')
                    self.record_result('Python Dependencies', 'PASS', 'All dependencies available')
                    
            except Exception as e:
                Logger.warning(f'Could not verify package installation: {e}')
                self.record_result('Python Dependencies', 'WARNING', 'Could not verify packages')
        else:
            Logger.info('No requirements.txt found (may be optional)')
            self.record_result('Python Dependencies', 'INFO', 'No requirements.txt')
    
    async def check_environment_files(self) -> None:
        """Check for environment configuration files"""
        Logger.header('Environment Files Check')
        
        env_files = [
            {'path': 'course-template/.env.example', 'required': False, 'description': 'Environment template'},
            {'path': 'course-template/.gitignore', 'required': True, 'description': 'Git ignore file'}
        ]
        
        all_good = True
        for env_file in env_files:
            if Path(env_file['path']).exists():
                Logger.success(f"{env_file['description']}: {env_file['path']}")
            elif env_file['required']:
                Logger.error(f"Missing required file: {env_file['path']}")
                all_good = False
            else:
                Logger.info(f"Optional file not found: {env_file['path']}")
        
        self.record_result('Environment Files', 'PASS' if all_good else 'FAIL',
                          'Environment setup acceptable' if all_good else 'Missing required files')
    
    async def check_optional_tools(self) -> None:
        """Check for optional development tools"""
        Logger.header('Optional Tools Check')
        
        optional_tools = [
            {'name': 'Node.js', 'command': ['node', '--version']},
            {'name': 'npm', 'command': ['npm', '--version']},
            {'name': 'Java', 'command': ['java', '-version']},
            {'name': 'Docker', 'command': ['docker', '--version']},
            {'name': 'curl', 'command': ['curl', '--version']},
            {'name': 'wget', 'command': ['wget', '--version']}
        ]
        
        for tool in optional_tools:
            try:
                subprocess.run(tool['command'], capture_output=True, check=True)
                Logger.success(f"{tool['name']}: Available")
            except (subprocess.CalledProcessError, FileNotFoundError):
                Logger.info(f"{tool['name']}: Not found (optional)")
        
        self.record_result('Optional Tools', 'INFO', 'Optional tools checked')
    
    async def check_network_access(self) -> None:
        """Check network connectivity to important services"""
        Logger.header('Network Access Check')
        
        test_urls = [
            'https://pypi.org/',
            'https://github.com',
            'https://api.github.com'
        ]
        
        network_issues = False
        for url in test_urls:
            try:
                # Use urllib for simple network check
                import urllib.request
                import urllib.error
                
                with urllib.request.urlopen(url, timeout=5) as response:
                    if response.status == 200:
                        Logger.success(f"Network access to {url}: OK")
                    else:
                        Logger.warning(f"Network access to {url}: Status {response.status}")
                        network_issues = True
            except (urllib.error.URLError, TimeoutError) as e:
                Logger.warning(f"Network access to {url}: Failed ({e})")
                network_issues = True
        
        self.record_result('Network Access', 'WARNING' if network_issues else 'PASS',
                          'Some network issues detected' if network_issues else 'Network access OK')
    
    async def check_file_permissions(self) -> None:
        """Check file system permissions"""
        Logger.header('File Permissions Check')
        
        try:
            # Test write permissions in current directory
            test_file = Path(f'.setup-test-{datetime.now().timestamp()}')
            test_file.write_text('test')
            test_file.unlink()
            
            Logger.success('File write permissions: OK')
            self.record_result('File Permissions', 'PASS', 'Write permissions available')
        except Exception as e:
            Logger.error(f"File permission issue: {e}")
            self.record_result('File Permissions', 'FAIL', f"Permission error: {e}")
    
    def record_result(self, category: str, status: str, details: str) -> None:
        """Record a verification result"""
        self.results.append({
            'category': category,
            'status': status,
            'details': details
        })
        
        if status == 'PASS':
            self.passed += 1
        elif status == 'FAIL':
            self.failed += 1
        elif status == 'WARNING':
            self.warnings += 1
    
    def generate_summary(self) -> None:
        """Generate and display verification summary"""
        Logger.header('Setup Verification Summary')
        
        print(f"{Colors.GREEN}✅ Passed: {self.passed}{Colors.RESET}")
        print(f"{Colors.RED}❌ Failed: {self.failed}{Colors.RESET}")
        print(f"{Colors.YELLOW}⚠️  Warnings: {self.warnings}{Colors.RESET}")
        
        if self.failed == 0:
            Logger.success('🎉 Congratulations! Your development environment is ready for the course.')
        else:
            Logger.error('⚠️  Please fix the failed checks before starting the course.')
        
        # Generate detailed report
        print('\n' + Colors.BOLD + 'Detailed Results:' + Colors.RESET)
        print('=' * 80)
        
        for result in self.results:
            status_color = (Colors.GREEN if result['status'] == 'PASS' else
                          Colors.RED if result['status'] == 'FAIL' else Colors.YELLOW)
            
            print(f"{status_color}{result['status']:<8}{Colors.RESET} "
                  f"{result['category']:<20} {result['details']}")
        
        # Save report to file
        self.save_report()
        
        print('\n' + Colors.CYAN + 'Next Steps:' + Colors.RESET)
        if self.failed > 0:
            print('1. Fix the failed checks listed above')
            print('2. Re-run this verification script')
            print('3. Contact your instructor if you need help')
        else:
            print('1. Start with Module 01 - Fundamentals')
            print('2. Read the course README.md for detailed instructions')
            print('3. Join the course discussion forum')
        
        print('\n' + Colors.BLUE + 'Support Resources:' + Colors.RESET)
        print('- Course Documentation: README.md')
        print('- Setup Guide: docs/setup.md')
        print('- Discussion Forum: [course-forum-url]')
        print('- Instructor Email: [instructor-email]')
    
    def save_report(self) -> None:
        """Save verification report to JSON file"""
        report = {
            'timestamp': datetime.now().isoformat(),
            'platform': f"{platform.system()} {platform.machine()}",
            'python_version': sys.version,
            'summary': {
                'passed': self.passed,
                'failed': self.failed,
                'warnings': self.warnings
            },
            'results': self.results
        }
        
        try:
            with open('setup-verification-report-python.json', 'w') as f:
                json.dump(report, f, indent=2)
            Logger.info('📄 Detailed report saved to: setup-verification-report-python.json')
        except Exception as e:
            Logger.warning(f'Could not save detailed report file: {e}')

async def main():
    """Main function to run verification"""
    verifier = SetupVerifier()
    
    try:
        await verifier.run_all_checks()
        
        # Exit with appropriate code
        sys.exit(1 if verifier.failed > 0 else 0)
    except Exception as e:
        Logger.error(f"Verification script error: {e}")
        sys.exit(1)

def show_help():
    """Show help message"""
    print("""
Course Template Setup Verification (Python)

Usage: python setup-verification.py [options]

Options:
  --help, -h     Show this help message
  --version, -v  Show version information
  --verbose      Enable verbose output
  --json         Output results in JSON format only

This script verifies that your Python development environment is properly
configured for the software development course.

Examples:
  python setup-verification.py
  python setup-verification.py --verbose
  python setup-verification.py --json > report.json
""")

if __name__ == '__main__':
    import asyncio
    
    # Handle command line arguments
    args = sys.argv[1:]
    
    if '--help' in args or '-h' in args:
        show_help()
        sys.exit(0)
    
    if '--version' in args or '-v' in args:
        print('Course Template Setup Verification (Python) v1.0.0')
        sys.exit(0)
    
    # Run async main function
    asyncio.run(main())